groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+// the variant header.
+#include <bslstl_monostate.h>
+
+#include <bslalg_swaputil.h>
+
+#include <bslma_constructionutil.h>
+#include <bslma_destructionutil.h>
+#include <bslma_stdallocator.h>
+#include <bslma_usesbslmaallocator.h>
+
+#include <bslmf_allocatorargt.h>
+
+#include <bslmf_integersequence.h>
+#include <bslmf_integralconstant.h>
+#include <bslmf_invokeresult.h>
+#include <bslmf_isbitwisemoveable.h>
+#include <bslmf_isconvertible.h>
+#include <bslmf_isreference.h>
+#include <bslmf_issame.h>
+#include <bslmf_makeintegersequence.h>
+#include <bslmf_movableref.h>
+#include <bslmf_nestedtraitdeclaration.h>
+#include <bslmf_removeconst.h>
+#include <bslmf_removecvref.h>
+#include <bslmf_util.h>
+#include <bslmf_voidtype.h>
+
+#include <bsls_assert.h>
+#include <bsls_compilerfeatures.h>
+#include <bsls_exceptionutil.h>
+#include <bsls_keyword.h>
+#include <bsls_libraryfeatures.h>
+#include <bsls_objectbuffer.h>
+#include <bsls_util.h>  // 'forward<T>(V)'
+
+#include <stddef.h>
+
+// We provide two set of bsl::variant feature sets. These are the necessary
+// features for the more functional set.
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_DECLTYPE
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_GENERALIZED_INITIALIZERS
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_RVALUE_REFERENCES
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+// Note that BSLS_LIBRARYFEATURES_HAS_CPP14_INTEGER_SEQUENCE does not require a
+// C++14 compiler
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP14_INTEGER_SEQUENCE
+#define BSL_VARIANT_FULL_IMPLEMENTATION
+#endif  // BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+#endif  // BSLS_LIBRARYFEATURES_HAS_CPP14_INTEGER_SEQUENCE
+#endif  // BSLS_COMPILERFEATURES_SUPPORT_RVALUE_REFERENCES
+#endif  // BSLS_COMPILERFEATURES_SUPPORT_GENERALIZED_INITIALIZERS
+#endif  // BSLS_COMPILERFEATURES_SUPPORT_DECLTYPE
+
+#ifdef BSL_VARIANT_FULL_IMPLEMENTATION
+#include <type_traits>
+// todo remove when replacing std::invoke with bsl::invoke
+#include <functional>
+#else  // BSL_VARIANT_FULL_IMPLEMENTATION
+#include <bslstl_function.h>
+#endif  // BSL_VARIANT_FULL_IMPLEMENTATION
+
+#ifdef BSLS_COMPILERFEATURES_SUPPORT_GENERALIZED_INITIALIZERS
+#include <initializer_list>
+#endif  // BSLS_COMPILERFEATURES_SUPPORT_GENERALIZED_INITIALIZERS
+
+#if BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+// Include version that can be compiled with C++03 Generated on Mon Oct  3
+// 16:56:46 2022 Command line: sim_cpp11_features.pl bslstl_variant.h
+#define COMPILING_BSLSTL_VARIANT_H
+#include <bslstl_variant_cpp03.h>
+#undef COMPILING_BSLSTL_VARIANT_H
+#else
+
+namespace BloombergLP {
+namespace bslma {
+
+// A type used as a default template type argument in places where we would get
+// an empty template argument set generated by the sim_cpp11_features.pl
+// script.
+#define VARIANT_NOT_A_TYPE void
+
+#if !BSLS_COMPILERFEATURES_SIMULATE_CPP11_FEATURES
+
+template <class TYPE = VARIANT_NOT_A_TYPE, class... OTHER>
+struct UsesBslmaAllocatorAny;
```

Comment:
It looks like it will compile, but it doesn't. At least it doesn't with my version of the script. The unused void argument serves as the extra template parameter in the primary template definition. That extra template parameter is needed to get the specialisation for the largest number of variadic arguments valid - otherwise the script generates X template arguments for the primary template, but there is T and the X template arguments in the specialisation.

Reply:
Ah that makes sense: you're saying that with my strategy, the primary template will have up to 10 template arguments, and one of the partial specializations will have `T` followed by 10 arguments, which will make it invalid. I think it would be helpful to elaborate on this a bit more in the comments.

Reply:
sure. I replaced the comment above `VARIANT_NOT_A_TYPE` with 
```// When generating the expansion of a variadic template, the script currently
// expands the primary template by the required number of template arguments.
// If there is a partial specialization that adds a non variadic template
// parameter, the generated primary template will no longer match the expansion
// of such specialization.  In the case of `variant` metafunction helpers, such
// additional template argument is always a type parameter.  To work around
// this issue, a dummy template type parameter is added to the primary template
// and defaulted to VARIANT_NOT_A_TYPE.
```

let me know if that works or if you want to move the comment to the metafunctions itself (and whether to just add it to the first one or to all the ones that use this trick).